docs/tsm/TSM_FORMAT.md

```diff
@@ -0,0 +1,161 @@
+# TSM Inverted Index Proposal
+
+This is a proposal for adding a disk-based, inverted index for measurements, tags and values.  This is a extension and/or alternative of the ideas proposed in [TSI Proposal #7174](https://github.com/influxdata/influxdb/pull/7174).
+
+The inverted index supports the following:
+
+* Hash lookups of measurement, tags, and values
+* Sorted iteration over measurement, tags and values
+* Efficient series boolean operations over compressed bitmaps
+* Compressed series keys
+
+The index is a separate set of files specific to a shard that are compacted and merged periodically.  Searches run across multiple files and are merged.
+
+## Format
+
+┌────────────────────────────────────────────────────────────────┐
+│                         Inverted Index                         │
+├────────────────────┬──────────────────┬────┬───────────────────┤
+│    Series Dict     │     Tag Sets     │... │   Measurements    │
+└────────────────────┴──────────────────┴────┴───────────────────┘
+
+The `Inverted Index` contains three sections: Series Dictionary, Tag Sets and Measurements.  The `Series Dictionary` contains a dictionary encoded version of all series in the index.  The `Tag Sets` are the unique tag/value pairs for a given measurement.  The `Measurements` are the unique set of measurements in the index.
+
+┌───────────────────────────────────────────────────────┐
+│                   Series Dictionary                   │
+├───────────────────────────┬───────────────────────────┤
+│        Dictionary         │          Series           │
+├─────────┬─────────────────┼─────────┬─────────────────┤
+│   Len   │  Sorted Terms   │   Len   │  Sorted Series  │
+│ 4 bytes │                 │ 4 bytes │                 │
+└─────────┴─────────────────┴─────────┴─────────────────┘
+
+The `Series Dictionary` contains two sections.  The first is an array of unique terms for all series keys.  For example, `cpu,host=server0`, would create an array `3cpu4host7server0` where each term is prefixed with it's variable byte encoded length.
+
+The series are a dictionary encoded version of the series keys.  Using the current example, it would be encoded as (0, 4, 9) where 0 is the offset where `cpu` begins, 4 is where `host` begins, and 9 is where `server0` begins.  These 3 small integers would be variable byte encoded as a byte string using the format:
+
+┌───────┬───────┬───────┬───────┐
+│ Meas  │  Tag  │  Tag  │ Value │
+│       │ Count │       │       │
+└───────┴───────┴───────┴───────┘
+
+`cpu,host=server0` would encode to 4 bytes.
```

Comment:
This looks great from a compression point-of-view (though to get 4 bytes per-series we would be saying that at most each series could have  255 tags on it I think so we can fit the tag count value in a  single byte?).

The trade-off here I guess will be the speed to decompress a series key. In the above example I think it would take three random disk seeks to get to the sorted terms, and then for each term-seek we need to do a variable-byte decode to get the term length, before reading it.

Could this bite us when dealing with the following requirement from the requirements doc?

> 10. The index must support finding all series by regex